### ISSUE TITLE
CloudInit configuration is not getting applied in a clone

### ISSUE THREAD

**Issue by some-random-username-123** [bpg/terraform-provider-proxmox#586]
**Describe the bug**
My issue is: When applying a terraform manifest that clones a vm and adds cloud init configuration, the cloudinit image doesnt get regenerated.
I the UI this is done via the Button Regererate image:
![Screenshot_20230922_165942](https://github.com/bpg/terraform-provider-proxmox/assets/145786212/e32d623c-0382-4ec0-a161-733550a34540)
There is also a API call that can do this: [API Refference](https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node}/qemu/{vmid}/cloudinit)

![Screenshot_20230922_170233](https://github.com/bpg/terraform-provider-proxmox/assets/145786212/dca9963a-f005-44b6-a044-d639fdbb9d5c)

**To Reproduce**
1. Create a resource with a initialization field
2. Run `terraform apply`
3. Wait until VM runs

My TF Module:
```terraform
...
variable "vm_ci_interface" {
    description = "The Interface for the Cloudinit drive on the VM."
    type = string
    default = "ide0"
}

variable "vm_template_id" {
    description = "The VM ID of the Template that will be cloned."
    type = number
    default = 9000
}

variable "vm_ip_address" {
    description = "The IP adress that the VM has. (use 'dhcp' for automatic IP)"
    type = string
    default = "dhcp"
}

variable "vm_ip_gateway" {
    description = "The IP of the gateway that the VM uses. (when vm_ip_address = 'dhcp' no gateway is used)"
    type = string
    default = ""
}

variable "vm_sa_user_name" {
    description = "The username of the service account that is used on this VM."
    type = string
    default = "sa_ansible"
}

variable "vm_sa_user_password" {
    description = "The password of the service account that is used on this VM."
    type = string
    sensitive = true
}

...
# --------
resource "proxmox_virtual_environment_vm" "almalinux9-basic" {
...
  initialization {
    interface       = var.vm_ci_interface

    ip_config {
      ipv4 {
        address     = var.vm_ip_address
        gateway     = var.vm_ip_address == "dhcp" ? null : var.vm_ip_gateway
      }
    }

    user_account {
      username      = var.vm_sa_user_name
      password      = var.vm_sa_user_password
      keys          = var.vm_sa_user_ssh_keys
    }
  }
......
```

**Screenshots**

**Additional context**
- Single or clustered Proxmox: Single
- Provider version (ideally it should be the latest version): latest
- Terraform version: latest
- OS (where you run Terraform from): Almalinux9 KDE

**Comment by bpg** [comment]
Hi @some-random-username-123! 👋🏼 (nice GH username btw :)

> 4. See error

Could you pls. post here the error that that you saw? 

Also I'd like to clarify whther the original VM that you're cloning has a cloudinit config?

**Comment by some-random-username-123** [comment]
Hi @bpg,
Sure I can give you more insights:

`Also I'd like to clarify whther the original VM that you're cloning has a cloudinit config?`
I Tried both, so i have a emplate with a Cloudinit drive, and i also have a template with no cloudinit drive.
I have the issue with both tamplates.

The exact issue is, that when I clone my vm with the TF provider, after the vm starts the cloudinit config (user and hostname, etc) is not applied in the template.

But when I clone it manually and press the button, all the cloudinit configuration is applied.


I hope this gives you some more insight.

Regards

**Comment by 0xinterface** [comment]
@bpg I took a brief look at the example @some-random-username-123 and apparently the packer template is built by specifying bios to ovmf therefore I think this is related to #575

**Comment by webtroter** [comment]
Hello,

I'm also having this problem

I found this small thread in the Proxmox forums about the Regeneration process. -> https://forum.proxmox.com/threads/regenerate-cloud-init-image-using-ansible.89964/

It is basically a remove/re-add the cloudinit drive.

I hope it could help solve this problem.

Thank you

**Comment by bpg** [comment]
Hey @some-random-username-123 👋🏼 

I'm trying to reproduce this issue with a minimal TF configuration, but everything seems to be working for me (tm)

Here is my template:
```hcl
 resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name = "586-template"

  node_name = var.virtual_environment_node_name
  vm_id     = 586
  machine   = "q35"
  bios      = "ovmf"

  agent {
    enabled = true
  }

  efi_disk {
    datastore_id = "local-lvm"
    file_format  = "raw"
    type         = "4m"
  }

  disk {
    datastore_id = "local-lvm"
    file_id      = proxmox_virtual_environment_file.ubuntu_cloud_image.id
    interface    = "virtio0"
    iothread     = true
    discard      = "on"
    size         = 20
  }

  network_device {
    bridge = "vmbr0"
  }

  operating_system {
    type = "l26"
  }

  started = false
}

resource "proxmox_virtual_environment_file" "ubuntu_cloud_image" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = var.virtual_environment_node_name

  source_file {
    path = "https://cloud-images.ubuntu.com/jammy/current/jammy-server-cloudimg-amd64.img"
  }
}

data "local_file" "ssh_public_key" {
  filename = "./id_rsa.pub"
}

resource "proxmox_virtual_environment_file" "cloud_config" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = var.virtual_environment_node_name

  source_raw {
    data = <<EOF
#cloud-config
runcmd:
    - apt update
    - apt install -y qemu-guest-agent net-tools
    - timedatectl set-timezone America/Toronto
    - systemctl enable qemu-guest-agent
    - systemctl start qemu-guest-agent
    - echo "done" > /tmp/vendor-cloud-init-done
    EOF

    file_name = "cloud-config.yaml"
  }
}

resource "proxmox_virtual_environment_vm" "ubuntu_vm_cloned" {
  depends_on = [proxmox_virtual_environment_vm.ubuntu_vm]
  name = "587-cloned"
  node_name = var.virtual_environment_node_name
  vm_id     = 587

  clone {
    vm_id = 586
  }

  initialization {
    #    interface = "scsi0"
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }

    user_account {
      username = "ubuntu"
      keys     = [trimspace(data.local_file.ssh_public_key.content)]
    }

    vendor_data_file_id = proxmox_virtual_environment_file.cloud_config.id
  }

  started = true
}
```

Both VMs got deployed, the "cloned" one has cloudinit:
<img width="813" alt="Screenshot 2023-10-02 at 9 22 01 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/5c3b9a41-a88d-427e-8b86-c98cd909adf6">

Without pressing "regenerate" there, I was able to login into the VM using configured public key:
<img width="847" alt="Screenshot 2023-10-02 at 9 29 32 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/d2e5893f-fa81-4a1d-8aa9-726136a76043">

So far, no ideas what could be wrong 🤔

**Comment by bpg** [comment]
@some-random-username-123 [this](https://github.com/some-random-username-123/Proxmox-setup/blob/2eacc4eb816018997f21833596c6867548dc0310/terraform/modules/proxmox/almalinux9-basic/vars.tf#L108) could be the reason. If you're not overriding `vm_ci_interface` variable at apply, the default `ide0` may not work in all configurations. I would recommend switching to `ide2` as per [documentation](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#initialization).

**Comment by some-random-username-123** [comment]
@bpg  Thanks for the response.
I tried it with the other interface and all, but it still wont work. This might be a issue with the distribution, as you areusing ubuntu and it works for you, idk. I will try your ubuntu template and clone next and see it has something to do with the linux distribution, alternatively I will also try and reinstall proxmox and see if this might fixes the issue.

**Comment by bpg** [comment]
Hi @webtroter, @tobiasehlert! 👋🏼

I see you upvoted this issue, do you have a similar problem in your deployments? If yes, would you mind sharing your configuration?

Thanks!

**Comment by webtroter** [comment]
On my part, after your other reply I seemed to made it work (see my
librenms IAC repo)

But I need to retry the provider to be certain or at least report what I
did wrong.

**Comment by tobiasehlert** [comment]
@bpg, well slightly the same.

Example terraform file:
```terraform
resource "proxmox_virtual_environment_file" "vm_cloud_init" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = var.proxmox_settings.node

  source_raw {
    file_name = "cloud-init.user-data.yml"
    data      = file("cloud-init/user-data.yml")
  }
}
```

Example cloud-init/user-data.yml file:
```yaml
#cloud-config
package_upgrade: true
packages:
  - qemu-guest-agent

timezone: Europe/Stockholm
users:
  - name: myusername
    groups: [adm, cdrom, dip, plugdev, lxd, sudo]
    lock-passwd: false
    sudo: ALL=(ALL) NOPASSWD:ALL
    shell: /bin/bash
    ssh_authorized_keys:
      - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAINyDI3c8l5Se4w0B2ejL36WJh/PdD5MKKsL9zX0a9sT

power_state:
  delay: now
  mode: reboot
  message: Rebooting after cloud-init completion
  condition: true
```

Proxmox GUI:
![image](https://github.com/bpg/terraform-provider-proxmox/assets/4417850/1b6f0f98-339b-46f7-ab1d-78f15df8cfff)

My expectation was to see my user that was provisioned and the ssh-key, but that's not shown.

I run version Proxmox version 8.0.3 and Terraform provider bpg/proxmox version 0.33.0.

**Comment by tobiasehlert** [comment]
Also (maybe related to cloud-init issue), the _proxmox_virtual_environment_file_ resource for the cloud-init file shown in [comment](https://github.com/bpg/terraform-provider-proxmox/issues/586#issuecomment-1759733335) returns following error to me when applying:
```plain
Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: no such file: "local:snippets/cloud-init.user-data.yml"
│ 
│   with proxmox_virtual_environment_file.vm_cloud_init,
│   on cloud_init.tf line 1, in resource "proxmox_virtual_environment_file" "vm_cloud_init":
│    1: resource "proxmox_virtual_environment_file" "vm_cloud_init" {
│ 
╵
```

**Comment by bpg** [comment]
@tobiasehlert could you pls check if `snippets` content type is enabled on the `local` storage? Could be another case of https://github.com/bpg/terraform-provider-proxmox/issues/370#issuecomment-1605266793

**Comment by ratiborusx** [comment]
I've spent a whole day yesterday and quite a bit of overtime experimenting with provider and testing cloud-init functionality specifically and here's my observations.

PART I
In short, CI (cloud-init) cloud-config works as expected at least in userdata part. I didn't test vendor/network/meta because i've got no use for it but i have no reason to believe it won't work.
As i already mentioned somewhere in other issues Proxmox executes cloud-init in somewhat "unconventional" way (at least i'm not used to it compared to how cloud providers deal with it) generating an ISO image for it with the data and mounting it in CD-ROM drive. It also allows to regenerate said image and reapply configuration it contains (for example add a new ssh-key for a user), which is not usually the case with the cloud stuff where cloud-config applied only once at the first boot (or more precisely some modules of CI).

I'm not sure if people are aware but things that Proxmox exposes in it's GUI for Cloud-Init is just a small subset of what cloud-config is capable of. These things are what provider can control via "initialization.dns", "initialization.ip_config" and "initialization.user_account" (btw @bpg there seem to be no way for provider to change "Upgrade packages" option as in GUI?):
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/f8b7ce9f-51d1-4180-8e89-ccc681bd58b1)

But that is in no way a full extent of what could be done via cloud-config and by using provider's arguments like "user_data_file_id" (or network/vendor/meta ones) we could inject our own custom cloud-init cloud-config.
My own mistake (and probably most people with this issue are doing the same) was that i expected to see changes in exposed parameters of cloud-init in Proxmox's GUI - it won't happen although after booting up all the config's parameters would be applied. 
I also do not fully understand that weird magic Proxmox does with cloud-init so it confuses me alot.
For example (and i think THIS is what provider does or emulates when we inject custom cloud-config via "proxmox_virtual_environment_file") i used a console utility "qm" to inject my custom cloud-config into an image and bake it in template afterwards. Documentation also mentions that you could dump cloud-init info contained inside an image but this somehow never worked for me (for CUSTOM cloud-config anyway, but it shows info you put in the GUI fields of cloud-init).
An example:

```
qm set
...
--cicustom [meta=<volume>] [,network=<volume>] [,user=<volume>] [,vendor=<volume>]
           cloud-init: Specify custom files to replace the automatically generated ones at start.
           

# freshly baked template with an injected cloud-config shows nothing for it
root@prox-srv1:~# qm cloudinit dump 1011 user
#cloud-config
hostname: template-ubuntu-22
manage_etc_hosts: true
fqdn: template-ubuntu-22.04
chpasswd:
  expire: False
users:
  - default
package_upgrade: true

# i try to inject cloud-config via console tool but it still shows nothing afterwads
root@prox-srv1:~# qm set 1011 --cicustom "user=local:snippets/userdata-proxmox.yml"
update VM 1011: -cicustom user=local:snippets/userdata-proxmox.yml

root@prox-srv1:~# qm cloudinit dump 1011 user
#cloud-config
hostname: template-ubuntu-22
manage_etc_hosts: true
fqdn: template-ubuntu-22.04
chpasswd:
  expire: False
users:
  - default
package_upgrade: true

# but if i change "User" field of Cloud-Init in Proxmox's GUI it detects changes
root@prox-srv1:~# qm cloudinit dump 1011 user
#cloud-config
hostname: template-ubuntu-22
manage_etc_hosts: true
fqdn: template-ubuntu-22.04
user: WOLOLO <<<< CHANGED!
chpasswd:
  expire: False
users:
  - default
package_upgrade: true
```
In the end it's very weird for me how Proxmox handles cloud-init and i'm not sure what exactly it shows via "qm cloudinit dump". Why it's only reflection of GUI's changes i've no clue...
Maybe some Proxmox guru could explain it or @bpg himself (at least on provider's part, how it handles all that stuff).
The only thing i want to reiterate again - if you inject custom cloud-config into your vm/template do not expect to see any changes in the Proxmox's GUI in Cloud-Init. It won't be show in "qm cloudinit dump" either.

PART II
That's content of my custom cloud-config:
```
ratiborus@HOMEWORLD:~/WORKSPACE/terraform_yandex/proxmox$ cat ../keys/userdata-proxmox.yml
#cloud-config
hostname: px-cloud-config-applied
users:
  - name: tf
    groups: adm,sudo
    shell: /bin/bash
    sudo: ['ALL=(ALL) NOPASSWD:ALL']
    ssh-authorized-keys:
      - ssh-rsa AAAABsomelongstring== tf

packages:
- qemu-guest-agent
package_upgrade: true

runcmd:
  - systemctl enable --now qemu-guest-agent
  - echo "done" > /tmp/user-cloud-init-done

final_message: |
  cloud-init has finished
  version: $version
  timestamp: $timestamp
  datasource: $datasource
  uptime: $uptime
  Ratiborus was here... 
```

And here's 3 different implementations for injecting custom cloud-init cloud-config for userdata, and all of them work:
- cloud-config is kept in a dedicated yaml file and "imported" inside resource directly as external file; probably they way i'd use it if not one small issue i'll mention a bit later...
- cloud-config is kept in a dedicated yaml file and "imported" inside resource as raw data via "file" function of terraform that renders content of an external file; my favourite way of all so far but there may be unintended consequences, i didn't test it much
- cloud-config is kept inline inside terraform manifest and "imported" inside resource as raw data; probably the worst implementation because you keep potentially changeable data inside infrastructure manifest, but may work in some cases where you don't use modules or need a fast way to test something
```
#==================#
#   CLOUD-CONFIG   #
#==================#

resource "proxmox_virtual_environment_file" "cloud_config_userdata" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "prox-srv1"
  source_file {
    path = "../keys/userdata-proxmox.yml"
  }
}

resource "proxmox_virtual_environment_file" "cloud_config_userdata_raw_file" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "prox-srv1"
  source_raw {
    data = file("../keys/userdata-proxmox.yml")
    file_name = "userdata-proxmox-raw-file.yml"
  }
}

resource "proxmox_virtual_environment_file" "cloud_config_userdata_raw" {
  content_type = "snippets"
  datastore_id = "local"
  node_name    = "prox-srv1"
  source_raw {
    data = <<EOF
#cloud-config
hostname: px-cloud-config-applied
packages:
  - qemu-guest-agent
users:
  - name: tf
    groups: sudo
    shell: /bin/bash
    ssh-authorized-keys:
      - ssh-rsa some_long_key
    sudo: ALL=(ALL) NOPASSWD:ALL
runcmd:
    - systemctl enable --now qemu-guest-agent
    - echo "done" > /tmp/user-cloud-init-done
EOF

    file_name = "userdata-proxmox-raw.yml"
  }
}
```

And now about the issue with the first implementation i mentioned earlier.
If you change your external file that is getting imported into Proxmox's snippets datastore these changes are not visible for provider and resource will stay the same. Most likely because provider checks only existence of a file by its name. But if we "import" raw source even as rendered external file then provider sees changes and acts accordingly. The same thing probably happens with cloud images being imported into ISO datastore.
Maybe if provider checked and compared checksum of source and target files every plan and tried to recreate resource (re-download file) if changes are detected it would work better. We could also re-download any distribution's cloud images when they get updated. But i'm not sure if it's something feasible or easily implemented or even wanted by most people.
What do you think, @bpg?

Here's an example:
```
module.proxmox.proxmox_virtual_environment_file.cloud_config_userdata_raw_file: Creating...
module.proxmox.proxmox_virtual_environment_file.cloud_config_userdata: Creating...
module.proxmox.proxmox_virtual_environment_file.cloud_config_userdata: Creation complete after 1s [id=local:snippets/userdata-proxmox.yml]
module.proxmox.proxmox_virtual_environment_file.cloud_config_userdata_raw_file: Creation complete after 1s [id=local:snippets/userdata-proxmox-raw-file.yml]

###
do some changes to source file "../keys/userdata-proxmox.yml", i added installation of vim and htop packages
###

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # module.proxmox.proxmox_virtual_environment_file.cloud_config_userdata_raw_file must be replaced
-/+ resource "proxmox_virtual_environment_file" "cloud_config_userdata_raw_file" {
      + file_modification_date = (known after apply)
      ~ file_name              = "userdata-proxmox-raw-file.yml" -> (known after apply)
      + file_size              = (known after apply)
      + file_tag               = (known after apply)
      ~ id                     = "local:snippets/userdata-proxmox-raw-file.yml" -> (known after apply)
        # (5 unchanged attributes hidden)

      ~ source_raw {
          ~ data      = <<-EOT # forces replacement
                #cloud-config
                hostname: px-cloud-config-applied
                users:
                  - name: tf
                    groups: adm,sudo
                    shell: /bin/bash
                    sudo: ['ALL=(ALL) NOPASSWD:ALL']
                    ssh-authorized-keys:
                      - ssh-rsa AAAABsomelongstring== tf

                packages:
                - qemu-guest-agent
              + - vim
              + - htop
                package_upgrade: true

                runcmd:
                  - systemctl enable --now qemu-guest-agent
                  - echo "done" > /tmp/user-cloud-init-done

                final_message: |
                  cloud-init has finished
                  version: $version
                  timestamp: $timestamp
                  datasource: $datasource
                  uptime: $uptime
                  Ratiborus was here...
            EOT
            # (2 unchanged attributes hidden)
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

**Comment by ratiborusx** [comment]
And another longread. I'm sorry...

@bpg
I'm not exactly sure what is the reason but it probably needs to be highlighted in documentation.
When you create a VM template do not forget to set "started = false". Default is "true" which is sane but not for template considering it's not supposed to be ran at all.
Else you'll be getting updates in-place on any plan generation even though VM itself wouldn't be started ever:

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.proxmox.proxmox_virtual_environment_vm.template_ubuntu_2204 will be updated in-place
  ~ resource "proxmox_virtual_environment_vm" "template_ubuntu_2204" {
        id                      = "1011"
      ~ ipv4_addresses          = [] -> (known after apply)
      ~ ipv6_addresses          = [] -> (known after apply)
        name                    = "template-ubuntu-22.04"
      ~ network_interface_names = [] -> (known after apply)
        tags                    = [
            "template-ubuntu-22.04",
            "templates",
            "terraform",
        ]
        # (21 unchanged attributes hidden)

        # (9 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```


Also, even with "started = false" i'm still getting this "architecture" update for template VM after it was already created (once on the next plan generation):

```
...
module.proxmox.proxmox_virtual_environment_vm.template_ubuntu_2204: Creating...
module.proxmox.proxmox_virtual_environment_vm.template_ubuntu_2204: Creation complete after 7s [id=1011]

ratiborus@HOMEWORLD:~/WORKSPACE/terraform_yandex/proxmox$ terraform apply
...

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.proxmox.proxmox_virtual_environment_vm.template_ubuntu_2204 will be updated in-place
  ~ resource "proxmox_virtual_environment_vm" "template_ubuntu_2204" {
        id                      = "1011"
        name                    = "template-ubuntu-22.04"
        tags                    = [
            "template-ubuntu-22.04",
            "templates",
            "terraform",
        ]
        # (24 unchanged attributes hidden)

      ~ cpu {
          + architecture = "x86_64"
            # (7 unchanged attributes hidden)
        }

        # (8 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

It happens only once after template VM resource creation and also does not happen if i recreate resource (i.e. with "terraform taint"). No difference whether "template = true" or "false".
I also tried it on other Linux distribution (Astra) and the result is the same. Here's test resources:

```
resource "proxmox_virtual_environment_file" "cloud_image_ubuntu_2204" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = "prox-srv1"

  source_file {
    path = "https://cloud-images.ubuntu.com/releases/22.04/release/ubuntu-22.04-server-cloudimg-amd64.img"
  }
}

resource "proxmox_virtual_environment_vm" "template_ubuntu_2204" {
  name        = "template-ubuntu-22.04"
  description = "Ubuntu 22.04 template"
  tags        = sort(["terraform", "templates", "template-ubuntu-22.04"])
  pool_id     = proxmox_virtual_environment_pool.templates.id

  node_name     = "prox-srv1"
  vm_id         = 1011
  migrate       = true
  on_boot       = false
  started       = false
  template      = true
  scsi_hardware = "virtio-scsi-single"

  agent {
    enabled = true
    trim    = true
  }

  cpu {
    architecture = "x86_64"
    cores        = 1
    numa         = true
    sockets      = 2
    type         = "host"
  }

  memory {
    dedicated    = 2048
  }

  disk {
    datastore_id = "local"
    file_id      = proxmox_virtual_environment_file.cloud_image_ubuntu_2204.id
    file_format  = "qcow2"
    interface    = "scsi0"
    cache        = "none"
    size         = 20
    discard      = "on"
    iothread     = true
    ssd          = true
  }

  initialization {
    datastore_id = "local"
    interface    = "ide2"
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }
    user_data_file_id = proxmox_virtual_environment_file.cloud_config_userdata.id
  }

  network_device {
    bridge = "vmbr0"
  }

  operating_system {
    type = "l26"
  }

  serial_device {
    device = "socket"
  }

  vga {
    enabled = true
    memory  = 32
    type    = "serial0"
  }

  # lifecycle {
  #   ignore_changes = [
  #     initialization[0].user_data_file_id
  #   ]
  # }
}
```

Not sure but maybe all that info should be covered in a dedicated issue? Just feel like i'm a bit off-topic here and deluting main question :)

**Comment by bpg** [comment]
Thanks for the wealth of information, @ratiborusx. That's a really impressive investigation! I'll need to set aside some time to go over all the details here. I'll get back to you at some point over the weekend :)

**Comment by ratiborusx** [comment]
Just a little bit of new information that was revealed at today's past-midnight meeting and demo.
If i inject my custom cloud-config like this there's no way to set any cloud-init configuration through Proxmox's GUI - it never gets applied even if i regenerate image:
```
  initialization {
    datastore_id = "local"
    interface    = "ide2"
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }
    user_data_file_id = proxmox_virtual_environment_file.cloud_config_userdata["prox-srv1"].id
  }
  ```
An example:
- I deployed VM with my custom cloud-config provided as above
- As part of that config i create user "tf" and drop its public ssh key
- Here's summary info about VM and its Cloud-Init status
  
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/6a4e69eb-9245-4326-852a-14e9923fb185)
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/a9f7f2fa-15ab-43d4-b940-a7423c851f97)

- Now i try to connect to that VM with my "tf" user and it's key:
```
ratiborus@HOMEWORLD:~$ ssh -i ~/.ssh/tf.key tf@10.177.144.35
tf@px-cloud-config-applied:~$ getent passwd | grep bash
root:x:0:0:root:/root:/bin/bash
tf:x:1000:1000::/home/tf:/bin/bash
```
- After that i change GUI's cloud-init config (added another user but with the same key as "tf"), press "Regenerate image" and power-cycle VM to be sure
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/0ecc9aa7-a862-44d3-a713-97df3e51e7fe)
- Trying to connect with new credentials:
```
ratiborus@HOMEWORLD:~$ ssh -i ~/.ssh/tf.key abc123@10.177.144.35
abc123@10.177.144.35: Permission denied (publickey).
```
- NUH-UH!

I saw those in resource description but thought they just overwrite each other.
So if after applying my custom cloud-config i reapply setting from GUI (with image regeneration) they will overwrite old ones from custom config because both are supposed to put that data in that small 4Mb cloud-init image that resides alongside with disk's one?
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/ae5343a1-6850-4b6b-a8d9-d785992000be)
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/ce5f4ac7-4185-4b1d-bfd3-0a14d0149d8b)

UPD.
Now, after reading some docs of PVE again i see that they call stuff available through GUI as "automatically generated config" and that custom and automatic stuff can be mixed but only on "silo" level - i.e. user, network, vendor and meta data should be unique (either custom or automatic).
![image](https://github.com/bpg/terraform-provider-proxmox/assets/123507924/31ba7f6f-7623-4009-aa60-c3110445d8ce)
After all that i'm still not sure whether i could use GUI's Cloud-Init settings to overwrite ones injected through custom cloud-config. It's not that i'd need that but others who access Proxmox via GUI would need it sometimes to drop in their own creds.
But ofc it has nothing to do with provider :)


@tobiasehlert
Your 1st issue is that you expect info from your custom cloud-config to be shown in GUI - it won't be visible there but it should be applied by cloud-init, your should check it.
Your 2nd issue with an error about missing file may be because you removed that file manually via GUI but resource still present in terraform's state file. The easiest way to fix would be to remove said resource from state file via "terraform state rm 'resource_name'" and recreate it. If I'm not wrong this issue (it also affected disk images) should be fixed in 0.36.0, i mentioned this in #643.

**Comment by tobiasehlert** [comment]
> @tobiasehlert could you pls check if `snippets` content type is enabled on the `local` storage? Could be another case of [#370 (comment)](https://github.com/bpg/terraform-provider-proxmox/issues/370#issuecomment-1605266793)

Yes, that fixed my problem with snippet, thanks @bpg!
New to Proxmox, so didn't know about that parameter :)

**Comment by Fabiosilvero** [comment]
Hello ! 

I also have this issue (I didn't wanted to open a duplicate, feel free to ask me to open my own issue).

Host : 
* Proxmox VE 8.0.4 (single node)
* OS : Debian 12 up to date

Template : 
* Built with Packer
* OS : Debian 12
* Cloud init installed
* OVMF

Terraform code (redacted) : 

```tf
terraform {
  required_providers {
    proxmox = {
      source = "bpg/proxmox"
      version = "0.37.0"
    }
  }
}

provider "proxmox" {
  endpoint = "https://my.awesome.server:8006/"
  api_token = "user@pve!token1=my-awesome-token"
  insecure = true
  tmp_dir  = "/var/tmp"
}

resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name        = "terraform-provider-proxmox-ubuntu-vm"
  description = "Managed by Terraform"
  tags        = ["terraform", "ubuntu"]

  node_name = "node1"
  vm_id     = 4321

  agent {
    # read 'Qemu guest agent' section, change to true only when ready
    enabled = true
  }

  clone {
    vm_id = "101"
  }

  bios = "ovmf"
  reboot = true
  boot_order = ["ide2","virtio0","net0"]

  disk {
    datastore_id = "datastore1"
    file_format  = "raw"
    interface    = "virtio0"
    size         = 25 
  }

  disk {
    datastore_id = "datastore1"
    file_format  = "raw"
    interface    = "virtio1"
    size         = 4 
  }

  network_device {
    bridge = "vmbr0"
    vlan_id = "1"
  }

  operating_system {
    type = "l26"
  }

  scsi_hardware = "virtio-scsi-single"

  initialization {
    datastore_id = "datastore1"
    interface = "ide2"
    ip_config {
        ipv4 {
            address = "1.2.3.4/24"
            gateway = "1.2.3.1"
        }
    }
  }

}

```
The VM is created from template with correct boot order, but after disk resize, the boot order is changed to `virtio0;net0` although I specified it in my Terraform code.

I tried to remove the Cloud Init Drive to let Terraform create it (Packer create this drive at ide0 at the last step) and set it to ide2, but I have the same issue.

**Comment by ratiborusx** [comment]
@Fabiosilvero
Hey!
I don't remember ever booting from Cloud-Init drive. It is my understanding that it only keeps cloud-config on it. You probably don't need it in the boot order at all, but i'm not sure if UEFI has any influence on the whole thing (never used "OVMF").

I don't think Terraform's provider will be able to save whatever you backed in your image about boot order, it will always rewrite it. Even if you don't specify it in the resource i think it'll default to "boot_order = []" which most likely means to boot up from the first drive it detects.
I'd recommend you to remove "ide2" from load order or try to comment out "boot order" line as a whole (so provider will switch to defaults) and check if it works.

Just to be sure i'd recommend you to try to create a template with an official cloud image of Debian 12 and see if the issue is image-related. I used this one and it works without any problems with cloud-config:
https://cloud.debian.org/images/cloud/bookworm/latest/debian-12-generic-amd64.qcow2

**Comment by Fabiosilvero** [comment]
Hello and thanks for your reply ! 

I can't get the cloud image to boot with OVMF even with the efi_disk fully configured on terraform side (I used the tar.xz one as qcow2 wasn't valid for the provider). I'll try again later but I don't think the template is the issue as it works perfectly with Telmate provider.

I get what you mean by removing boot_order, that was a workaround since with OVMF, it seems that if disks aren't present in boot_order, they aren't properly recognized (see https://forum.proxmox.com/threads/cloud-init-drive-not-auto-regenerating.117318/ and https://forum.proxmox.com/threads/ide-cloud-init-volumes-are-not-usable-when-cold-booting-a-vm.117345/).

I tried first without setting it, and the result was the same. My configuration quoted was misleading, that was a fluke but I tried with virtio0 first (and ide2 at the end) and I tried without ide2 : still same thing.

I don't think that this provider enforce the order in the Terraform config at the very end of the execution, but before any disks resize - and the boot order is altered at that moment again, causing config revert to default (that's just a guess).

Maybe it's a "Regenerate Image" trigger issue in this provider : if I stop the VM then click on "Regenerate Image" on the Web GUI, the Cloud Init drive appears in Boot Order and the Cloud Init service kicks in at startup.

**Comment by otopetrik** [comment]
Combination `cloud-init` + `ovmf` + `cloud-init drive @ ide` does not work,
combination `cloud-init` + `ovmf` + `cloud-init drive @ sata` also does not work,
but combination `cloud-init` + `ovmf` + `cloud-init drive @ scsi` **works** (debian 12 genericcloud image).
Not sure why.

Cloud-init [docs](https://cloudinit.readthedocs.io/en/latest/reference/datasources/nocloud.html) do not mention anything about different requirements when booted using UEFI.
Maybe OVMF does not support CD drive on `ide` or `sata` interfaces ?

Added 2 lines to a debian [example](https://gist.github.com/otopetrik/f6e6ef79636592cab957407da1ac5f7b):
```
...
bios = "ovmf" # added
...
initialization {
  interface = "scsi0" # added
  ..
}
```
and created the VM. Cloud-init correctly applied the configuration.

It is likely unsafe to modify the provider to change `initialization.interface` default to depend on `bios` value.
Any `scsiN` value could potentially colide with actual VM disks.
But it might be a good idea to detect when the user is using `bios = "ovmf"` with non-scsi interface, and display a warning/suggestion to use one of`scsi0`, ..., `scsi30` values. (and also add it to the documentation)

**Comment by some-random-username-123** [comment]
Hey again,

After a long time of testing around with different configs, I still got no results.

But now I tried the Config from @otopetrik (thank you btw) it works. 

When i use scsi1 [here](https://github.com/some-random-username-123/Proxmox-setup/blob/main/terraform/proxmox/tsttf01.tf#L15), all the CI config is getting applied to the VM on first boot.

So thanks to all of you that helped.

**Comment by Fabiosilvero** [comment]
Well, thank you ! I just changed the interface from ide0/2 to scsi1, and it worked flawlessly in OVMF setting.

Maybe a VM with OVMF requires that IDE drives should be present at boot order to be available in VMs at all, and SCSI drives don't need that (I don't see the Cloud Init drive in Boot order).

**Comment by windowsrefund** [comment]
I've learned some things about Proxmox+cloud-init which may add some value to some areas of this thread. 

- The Proxmox GUI only shows and interacts with a tiny subset of what is possible with cloud-init.
- In order to build a custom cloud-init, all 4 types must be passed to --cicustom as identified on this [thread](https://forum.proxmox.com/threads/cloud-init-network-configuration-with-ubuntu-cicustom-option.115746/#post-500438).
- When attempting to use `qm cloudinit dump` to view one or more types, only that small subsit which is "used" by the Proxmox GUI is shown. In other words, the `dump` approach is useless when working with a custom cloud-init.
- The dump approach mentioned previously also can not report against the **vendor** type. That said, it is unknown to me if Proxmox supports it despite it needing to be represented when setting `--cicustom`.

In general, the clobbering nature of `proxmox_virtual_environment_vm.initialization.ip_config` and `proxmox_virtual_environment_vm.initialization.user_account` over their *_data_file_id counterparts seems really finicky and just ripe for pointless confusion. I'm personally going to avoid using them and just stick with the data files since that's where all the cloud-init power lies anyway. I understand the value in the perception end users are quickly bootstrapping all the things but wonder if that's enough reason to continue supporting these conflicting elements. Food for thought.

**Comment by ratiborusx** [comment]
@windowsrefund 
Yep, that's basically what i wrote somewhere at the beginning of the thread. Not as concise though, i must admit :)
1. You don't need to pass all types of cloud-config (meaning user/network/vendor/meta).
2. Those elements (meaning GUI cloud-config subset and all the custom stuff) are not conflicting on the silo/type level so to say. Meaning you can pass userdata via GUI and some other data via custom config. 
3. That's basically what i came to - i pass custom cloud-config userdata but leave possibility for end users to pass network configuration via either "initialization" arguments in provider or GUI.
4. I'm still not sure if it's possible to overwrite custom cloud-config with the stuff you pass via GUI. For now all my research points to resounding no. That's actually was a major bummer for me - i wanted to preserve possibility for end users to change userdata via GUI, so they could for example add their own ssh-keys to the guest VM. I just gave up on that i guess.
5. Yes, you can pass custom cloud-config vendordata too - i think if you check examples in github you'll find it. But keep in mind that in case there's the same cloud-init modules used in user and vendor datas they are merged. And userdata takes precedence, meaning vendordata gets overwritten.

**Comment by windowsrefund** [comment]
@ratiborusx Yea, it was definitely your earlier comments that perked my interest and got me heading down the rabbit hole. I'm glad I did as I have a better understanding of things. I really had to spend a few hours doing things by hand (via qm) as well as mounting the read-only cloudinit itself in order to cat the files inside. Thanks for helping to shed the light.

**Comment by bpg** [comment]
Awesome work testing all of these! ❤️ If anyone is up for the challenge to summarize cloud init testing in docs, I'll greatly appreciate a PR!
If not, I'll try to find some time in the next few weeks to test and document these use cases, with examples.

**Comment by bpg** [comment]
@all-contributors please add @Fabiosilvero, @windowsrefund  for testing

**Comment by bpg** [comment]
@all-contributors please add @Fabiosilvero and @windowsrefund for testing

**Comment by bpg** [comment]
@all-contributors please add @windowsrefund for testing

**Comment by M0NsTeRRR** [comment]
Hello guys,

Cloud init with IDE2 (default) is working on my side, you can find the attached apply i've posted in an other issue https://github.com/bpg/terraform-provider-proxmox/issues/734#issuecomment-1826280294
You can find my terraform config [here](https://github.com/M0NsTeRRR/homelabv3-infra/blob/main/terraform/modules/vm/main.tf) (I use terragrunt to boostrap everything)

Tested with Proxmox Virtual Environment 8.0.9, Terraform v1.6.4 and Ubuntu 22.04.3.

I'm creating a template with packer, with some ansible scripts (in the same repo), at the end I reset cloud init instance with `cloud-init clean --seed --machine-id` and I deploy this template with Terraform.

Note : I use cloud init file
```
  initialization {
    interface            = "ide2"
    network_data_file_id = proxmox_virtual_environment_file.cloud_network_config.id
    user_data_file_id    = proxmox_virtual_environment_file.cloud_user_config.id
    meta_data_file_id    = proxmox_virtual_environment_file.cloud_meta_config.id
  }
```

**Comment by cursedforever** [comment]
Hello guys, I don't really understand why the cloud-init configuration files are split in this way? Can't we just put one cloud-init file and then --cicustom in the backend? The way it's handled here seems cumbersome to me

**Comment by bpg** [comment]
Hi @cursedforever! That's what the Proxmox API [provides](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#qm_cloud_init) us: 

>10.8.3. Custom Cloud-Init Configuration
>
>The Cloud-Init integration also allows custom config files to be used instead of the automatically generated configs. This is done via the cicustom option on the command line:
>
> ```qm set 9000 --cicustom "user=<volume>,network=<volume>,meta=<volume>"```
>
>The custom config files have to be on a storage that supports snippets and have to be available on all nodes the VM is going to be migrated to. Otherwise the VM won’t be able to start. For example:
>
> ```qm set 9000 --cicustom "user=local:snippets/userconfig.yaml"```
>
> There are three kinds of configs for Cloud-Init. The first one is the `user` config as seen in the example above. The second is the `network` config and the third the `meta` config. They can all be specified together or mixed and matched however needed. The automatically generated config will be used for any that don’t have a custom config file specified.

Well, they missed `vendor` config in this section, but mentioned it a bit further in the doc.

I'm not a cloud-init expert by any means, so can't comment on "why" it's done this way. Tough I did try passing user config in `vendor` attribute, and it seemed to be working, by YMMV, and I suspect it may depend on the distro you're using.

**Comment by cursedforever** [comment]
Ok thank you very much, I've tried to make a global cloud-init file and pass it to the user_config_file variable and it works.

**Comment by michaelfranzl** [comment]
Using v0.49.0 with Proxmox 8.1.4 (with a VM with `bios = "ovmf"`), I am experiencing the same issue as described in the very first comment.

Adding `initialization.interface = "scsi1"` as suggested by @otopetrik [above](https://github.com/bpg/terraform-provider-proxmox/issues/586#issuecomment-1793602323) fixes the issue. Unfortunately, doing so also exposes issue https://github.com/bpg/terraform-provider-proxmox/issues/1083 where the resource is rebooted every _apply_ even when nothing changed.

**Comment by michaelfranzl** [comment]
Just for reference, more Proxmox users seem to have related issues with the cloud-init IDE interface, where switching to SCSI fixes the issue:

https://forum.proxmox.com/threads/cloud-init-image-only-applies-configuration-on-second-boot.93414/post-452830
https://forum.proxmox.com/threads/scsi0-cloud-init-drive-is-already-attached-at-ide0.131159/post-590571

Maybe the configuration default should be changed from IDE to SCSI?

**Comment by ratiborusx** [comment]
I personally never had problems with IDE interface for cloud-init but i must admit i never used UEFI ('bios = ovmf') either because i didn't need PCI devices pass-through so far.
As a workaround for #1083 you may try to use something like this (i think it's unlikely someone would want to change these settings on already deployed instance):
```hcl
lifecycle {
    ignore_changes = [
      initialization[0].datastore_id,
      initialization[0].interface
    ]
  }
```

**Comment by ElForastero** [comment]
I'm using 0.64.0 and the issue is still there. Just writing this to prevent it from being closed as stale.

Default cloud-image configuration from snippets isn't applied.

**Comment by bpg** [comment]
Hey @ElForastero 👋🏼  
As you can see from the 40+ comments, this issue is quite complex, and many factors could be affecting your specific setup.

The bottom line is that, in most cases discussed above, the problem wasn't with the provider but with VM configuration, cloud-init settings, the base OS image, etc. I'm keeping this issue open mainly to [summarize](https://github.com/bpg/terraform-provider-proxmox/issues/586#issuecomment-1793866618) those aspects in the documentation.

I suggest opening a new issue and explaining your use case so we can investigate and offer guidance.

Thanks!

**Comment by bpg** [comment]
I think this was fixed in #1885

Can't reproduce anymore. My config:
```hcl
resource "proxmox_virtual_environment_vm" "template" {
  name      = "template"
  node_name = "pve"
  machine   = "q35"
  started   = false
  template  = true

  disk {
    datastore_id = "local-lvm"
    file_id      = proxmox_virtual_environment_download_file.ubuntu_cloud_image.id
    interface    = "virtio0"
    iothread     = true
    discard      = "on"
    size         = 20
  }

  network_device {
    bridge = "vmbr0"
  }

  vga {
    type = "virtio"
  }
}

resource "proxmox_virtual_environment_download_file" "ubuntu_cloud_image" {
  content_type = "iso"
  datastore_id = "local"
  node_name    = "pve"

  url = "http://devfs/jammy/current/jammy-server-cloudimg-amd64.img"
}
```
```
❯ tofu apply -auto-approve
OpenTofu used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

OpenTofu will perform the following actions:

  # proxmox_virtual_environment_download_file.ubuntu_cloud_image will be created
  + resource "proxmox_virtual_environment_download_file" "ubuntu_cloud_image" {
      + content_type        = "iso"
      + datastore_id        = "local"
      + file_name           = (known after apply)
      + id                  = (known after apply)
      + node_name           = "pve"
      + overwrite           = true
      + overwrite_unmanaged = false
      + size                = (known after apply)
      + upload_timeout      = 600
      + url                 = "http://devfs/jammy/current/jammy-server-cloudimg-amd64.img"
      + verify              = true
    }

  # proxmox_virtual_environment_vm.clone will be created
  + resource "proxmox_virtual_environment_vm" "clone" {
      + acpi                    = true
      + bios                    = "seabios"
      + id                      = (known after apply)
      + ipv4_addresses          = (known after apply)
      + ipv6_addresses          = (known after apply)
      + keyboard_layout         = "en-us"
      + mac_addresses           = (known after apply)
      + migrate                 = false
      + name                    = "clone"
      + network_interface_names = (known after apply)
      + node_name               = "pve"
      + on_boot                 = true
      + protection              = false
      + reboot                  = false
      + reboot_after_update     = true
      + scsi_hardware           = "virtio-scsi-pci"
      + started                 = true
      + stop_on_destroy         = true
      + tablet_device           = true
      + template                = false
      + timeout_clone           = 1800
      + timeout_create          = 1800
      + timeout_migrate         = 1800
      + timeout_move_disk       = 1800
      + timeout_reboot          = 1800
      + timeout_shutdown_vm     = 1800
      + timeout_start_vm        = 1800
      + timeout_stop_vm         = 300
      + vm_id                   = (known after apply)

      + clone {
          + full    = true
          + retries = 1
          + vm_id   = (known after apply)
        }

      + initialization {
          + datastore_id = "local-lvm"

          + dns {
              + domain  = "example.com"
              + servers = [
                  + "172.16.0.15",
                  + "172.16.0.16",
                ]
            }

          + ip_config {
              + ipv4 {
                  + address = "192.168.3.233/24"
                  + gateway = "192.168.3.1"
                }
            }

          + user_account {
              + password = (sensitive value)
              + username = "ubuntu"
            }
        }
    }

  # proxmox_virtual_environment_vm.template will be created
  + resource "proxmox_virtual_environment_vm" "template" {
      + acpi                    = true
      + bios                    = "seabios"
      + id                      = (known after apply)
      + ipv4_addresses          = (known after apply)
      + ipv6_addresses          = (known after apply)
      + keyboard_layout         = "en-us"
      + mac_addresses           = (known after apply)
      + machine                 = "q35"
      + migrate                 = false
      + name                    = "template"
      + network_interface_names = (known after apply)
      + node_name               = "pve"
      + on_boot                 = true
      + protection              = false
      + reboot                  = false
      + reboot_after_update     = true
      + scsi_hardware           = "virtio-scsi-pci"
      + stop_on_destroy         = false
      + tablet_device           = true
      + template                = true
      + timeout_clone           = 1800
      + timeout_create          = 1800
      + timeout_migrate         = 1800
      + timeout_move_disk       = 1800
      + timeout_reboot          = 1800
      + timeout_shutdown_vm     = 1800
      + timeout_start_vm        = 1800
      + timeout_stop_vm         = 300
      + vm_id                   = (known after apply)

      + disk {
          + aio               = "io_uring"
          + backup            = true
          + cache             = "none"
          + datastore_id      = "local-lvm"
          + discard           = "on"
          + file_format       = (known after apply)
          + file_id           = (known after apply)
          + interface         = "virtio0"
          + iothread          = true
          + path_in_datastore = (known after apply)
          + replicate         = true
          + size              = 20
          + ssd               = false
        }

      + network_device {
          + bridge      = "vmbr0"
          + enabled     = true
          + firewall    = false
          + mac_address = (known after apply)
          + model       = "virtio"
          + mtu         = 0
          + queues      = 0
          + rate_limit  = 0
          + vlan_id     = 0
        }

      + vga {
          + memory = 16
          + type   = "virtio"
        }
    }

Plan: 3 to add, 0 to change, 0 to destroy.
proxmox_virtual_environment_download_file.ubuntu_cloud_image: Creating...
proxmox_virtual_environment_download_file.ubuntu_cloud_image: Creation complete after 2s [id=local:iso/jammy-server-cloudimg-amd64.img]
proxmox_virtual_environment_vm.template: Creating...
proxmox_virtual_environment_vm.template: Still creating... [10s elapsed]
proxmox_virtual_environment_vm.template: Creation complete after 15s [id=100]
proxmox_virtual_environment_vm.clone: Creating...
proxmox_virtual_environment_vm.clone: Still creating... [10s elapsed]
proxmox_virtual_environment_vm.clone: Creation complete after 19s [id=101]

Apply complete! Resources: 3 added, 0 changed, 0 destroyed.
```

<img width="609" alt="Image" src="https://github.com/user-attachments/assets/c69b6ebc-49c7-4cec-acdc-f3db820d27ab" />

Logged-in as with `ubuntu` user/password from cloud-init:

<img width="710" alt="Image" src="https://github.com/user-attachments/assets/ef31446a-c662-44b4-9c9b-aafad977df8f" />
